src/extension/src/TelemetryWriter.py

```diff
@@ -13,75 +13,107 @@
-        raise NotImplementedError
+    @staticmethod
+    def __ensure_message_restriction_compliance(full_message):
+        """ Removes line breaks, tabs and restricts message to a character limit """
+        message_size_limit = Constants.TELEMETRY_MSG_SIZE_LIMIT_IN_CHARACTERS
+        formatted_message = re.sub(r"\s+", " ", str(full_message))
+        return formatted_message[:message_size_limit - 3] + '...' if len(formatted_message) > message_size_limit else formatted_message
+
+    def write_event(self, task_name, message, event_level=Constants.TelemetryEventLevel.Informational):
+        # create event json
+        # check if event size complies with restriction i.e. 6k
+        # write event to events file
+        if os.path.getsize(self.events_folder_path) >= Constants.TELEMETRY_DIR_SIZE_LIMIT_IN_CHARACTERS:
+            self.__delete_older_events()
+
+        event = self.__new_event_json(task_name, event_level, message)
+        if len(json.dumps(event)) > Constants.TELEMETRY_EVENT_SIZE_LIMIT_IN_CHARACTERS:
+            print("Cannot send data to telemetry as it exceeded the acceptable data size. [Data not sent={0}]".format(json.dumps(message)))
+        else:
+            self.write_using_temp_file(self.events_folder_path, event)
+
+    def __delete_older_events(self):
+        """ Delete older events until the atleast one new event file can be added as per the size restrictions """
+        try:
+            if os.path.getsize(self.events_folder_path) < Constants.TELEMETRY_DIR_SIZE_LIMIT_IN_CHARACTERS - Constants.TELEMETRY_EVENT_FILE_SIZE_LIMIT_IN_CHARACTERS:
+                print("Not deleting any existing event files as the event directory does not exceed max limit. At least one new event file can be added.")
+                return
+
+            print("Events directory size exceeds maximum limit. Deleting older event files until at least one new event file can be added.")
+            event_files = [os.path.join(self.events_folder_path, event_file) for event_file in os.listdir(self.events_folder_path) if (event_file.lower().endswith(".json"))]
+            event_files.sort(key=os.path.getmtime, reverse=True)
+
+            for event_file in event_files:
+                try:
+                    if os.path.getsize(self.events_folder_path) < Constants.TELEMETRY_DIR_SIZE_LIMIT_IN_CHARACTERS - Constants.TELEMETRY_EVENT_FILE_SIZE_LIMIT_IN_CHARACTERS:
+                        print("Not deleting any more event files as the event directory has sufficient space to add at least one new event file")
+                        break
+
+                    if os.path.exists(event_file):
+                        os.remove(event_file)
+                        print("Deleted [File={0}]".format(repr(event_file)))
+                except Exception as e:
+                    print("Error deleting event file. [File={0}] [Exception={1}]".format(repr(event_file), repr(e)))
```

Comment:
Since this is swallowing exceptions, what's the recovery/continuation strategy if there are rampant failures (like all event files can't be deleted)? At what point is it a terminal exception?

Reply:
Changed to throw a terminal exception is no files were deleted or the directory size exceeds limit even after deletion attempts